launcher/src/main/java/org/apache/spark/launcher/Main.java

```diff
@@ -0,0 +1,106 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Command line interface for the Spark launcher. Used internally by Spark scripts.
+ */
+public class Main extends LauncherCommon {
+
+  /**
+   * Usage: Main [class] [class args]
+   * <p/>
+   * This CLI works in two different modes:
+   * <ul>
+   *   <li>"spark-submit": if <i>class</i> is "org.apache.spark.deploy.SparkSubmit", the
+   *   {@link SparkLauncher} class is used to launch a Spark application.</li>
+   *   <li>"spark-class": if another class is provided, an internal Spark class is run.</li>
+   * </ul>
+   *
+   * The ultimate command will not be run in the same process. Instead, the command to be executed
+   * will be printed to stdout. On Unix systems, this will be one argument per line. On Windows
+   * systems, this will be a single line containing the command to be executed.
+   */
+  public static void main(String[] argsArray) throws Exception {
+    checkArgument(argsArray.length > 0, "Not enough arguments: missing class name.");
+
+    List<String> args = new ArrayList<String>(Arrays.asList(argsArray));
+    String className = args.remove(0);
+
+    boolean printLaunchCommand = false;
+    AbstractLauncher<?> launcher;
+    try {
+      if (className.equals("org.apache.spark.deploy.SparkSubmit")) {
+        launcher = new SparkSubmitCliLauncher(args);
+        printLaunchCommand = !isEmpty(System.getenv("SPARK_PRINT_LAUNCH_COMMAND"));
+      } else if (className.equals("pyspark")) {
+        launcher = new PySparkLauncher(args);
+      } else {
+        launcher = new SparkClassLauncher(className, args);
+        printLaunchCommand = !isEmpty(System.getenv("SPARK_PRINT_LAUNCH_COMMAND"));
+      }
+    } catch (IllegalArgumentException e) {
+      launcher = new UsageLauncher();
+    }
+
+    List<String> cmd = launcher.buildLauncherCommand();
+    if (printLaunchCommand) {
+      System.err.println("Spark Command: " + join(" ", cmd));
+      System.err.println("========================================");
+    }
+
+    for (String c : cmd) {
+      System.out.println(c);
+    }
```

Comment:
We should add a comment here explaining that the output here is consumed by a script that runs these as a command. Also, here we use the new line character as a delimiter. What happens if the command itself contains new lines, are they escaped properly?